File: src/extract_logs.py
import sys
import os

def extract_logs(date, log_file, output_dir="output"):
    """
    Extract logs for a specific date from a large log file.

    Args:
        date (str): The date to filter logs (YYYY-MM-DD).
        log_file (str): Path to the large log file.
        output_dir (str): Directory to save the output file.
    """
    if not os.path.exists(log_file):
        print(f"Error: Log file '{log_file}' does not exist.")
        return

    # Ensure output directory exists
    os.makedirs(output_dir, exist_ok=True)

    output_file = os.path.join(output_dir, f"output_{date}.txt")
    
    try:
        with open(log_file, 'r') as file, open(output_file, 'w') as output:
            for line in file:
                # Match lines starting with the given date (up to the 'T')
                if line.startswith(date + "T"):
                    output.write(line)
        print(f"Logs for {date} have been saved to {output_file}.")
    except Exception as e:
        print(f"An error occurred: {e}")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python extract_logs.py <YYYY-MM-DD>")
        sys.exit(1)
    
    # Input date from command-line argument
    date = sys.argv[1]
    log_file = "logs_2024.log"  # Adjust the file name as needed
    
    extract_logs(date, log_file)

File: src/extract_logs.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>

using namespace std;

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " YYYY-MM-DD" << endl;
        return 1;
    }

    string date = argv[1];  

    string logFile = "logs_2024.log"; 

    ifstream file(logFile);
    if (!file) {
        cerr << "Error: Could not open log file '" << logFile << "'" << endl;
        return 1;
    }

    string outputFile = "output/output_" + date + ".txt";
    ofstream outFile(outputFile);
    if (!outFile) {
        cerr << "Error: Could not create output file '" << outputFile << "'" << endl;
        return 1;
    }

    string line;
    while (getline(file, line)) {
        if (line.rfind(date, 0) == 0) { 
            outFile << line << '\n';
        }
    }

    cout << "Logs for " << date << " extracted successfully in: " << outputFile << endl;

    file.close();
    outFile.close();
    return 0;
}
